### PR TITLE
everything

### PR DESCRIPTION
**Pull request by davidchambers**
Once this is published we'll be able to define `S.lens`, `S.view`, `S.over`, and `S.set`. :)

I won't describe `Constant a b` here as the documentation should explain its use (although not necessarily its purpose). If you believe the documentation is lacking in any way, please say so.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+      jsc.constant (Array),
+      ConstantArb (String) (jsc.string),
+    ],
+    composition: [
+      jsc.constant (Array),
+      jsc.constant (Identity),
+      ConstantArb (Array) (jsc.array (IdentityArb (jsc.string))),
+    ],
+  });
+});
```
Comment (davidchambers) [inline test/index.js:331]: @Avaq, please pay particularly close attention to the traversable laws (which I still don't fully understand).

File: test/index.js [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    eq (show (Constant (Array) (['foo', 'bar', 'baz'])))
+       ('Constant (Array) (["foo", "bar", "baz"])');
+    eq (show (Constant (Constant (Constant (Number)))
+                       (Constant (Constant (Number))
+                                 (Constant (Number)
+                                           (-0)))))
+       ('Constant (Constant (Constant (Number)))' +
+                ' (Constant (Constant (Number))' +
+                          ' (Constant (Number)' +
+                                    ' (-0)))');
```
Comment (davidchambers) [inline test/index.js]: :joy:

File: index.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    //. ```javascript
+    //. > S.of (Constant (Array)) (42)
+    //. Constant (Array) ([])
+    //.
+    //. > S.of (Constant (String)) (42)
+    //. Constant (String) ('')
+    //. ```
+    (function() {
+      var empty;
+      try { empty = Z.empty (A); } catch (err) { return; }
```
Comment (masaeedu) [inline index.js]: What kinds of errors can happen here? Could the attachment of `['fantasy-land/of']` be done in the `Constant$bound` function with something like `Z.Monoid.test`?

Comment (masaeedu) [reply index.js]: Ah, I guess it can't be done in `Constant$bound` itself, since you need it before you have a value, but couldn't it nevertheless use `Z.Monoid.test(A)`?

Comment (davidchambers) [reply index.js]: `Z.Monoid.test` takes a value rather than a type representative:

```javascript
> Z.Monoid.test ('abc')
true

> Z.Monoid.test (String)
false
```

> What kinds of errors can happen here?

```javascript
> Z.empty (Number)
TypeError: Monoid.methods.empty(...) is not a function
```

I would like to avoid `try...catch`, but I can't see how to do so.

Comment (masaeedu) [reply index.js]: I guess it's off topic here, but do you think it would be a good idea to add an equivalent that does work on the type rep? Then if you have some function that can take a value and obtain it's type rep, and a function that can take a type rep and determine whether it implements a typeclass, their composition is `test`.

Comment (masaeedu) [reply index.js]: That would for example allow you to avoid using the `try`/`catch` here; all the class membership testing can be done on the `A` typerep outside of the `Constant$bound` constructor, and then in `Constant$bound` you just need to test whether the typerep of `value` is indeed `A`.

Comment (davidchambers) [reply index.js]: Great idea! I'll investigate changes to sanctuary-type-classes which would enable this.

Comment (masaeedu) [reply index.js]: Not 100% sure, but this approach might also yield some minor performance improvements elsewhere, by moving some checking out of the path where values are being moved around.

File: test/index.js (hunk 1 above)
Comment (Avaq) [reply test/index.js:331]: I'm not too familiar with the Traversable type class myself, as none of the types I maintain implement it. From what I can tell, nothing seems off about this test.

What does stand out to me in each of these tests is the level of specialization you're testing with, loosely indicated by how often a `jsc.constant` is used in the test parameters.

Ideally, every test would pass with an `AnyConstantArb` defined in terms of itself to include nesting (because the fantasy laws are often broken once nesting comes into play):

```js
const {AnyConstantArb} = jsc.letrec(tie => ({
  AnyConstantArb: jsc.oneof(
    ConstantArb(Number)(jsc.number),
    ConstantArb(String)(jsc.string),
    //...some more?
    ConstantArb(Constant)(tie('AnyConstantArb'))
  )
}))
```

We can also define an arbitrary function `a -> b`:

```js
const ABArb = jsc.oneof(
  a => ({a}),
  b => ({b}),
  c => [c],
  d => Identity['fantasy-land/of'](d)
);
```

Now the Functor composition parameters, for example, could be replaced with:

```js
[AnyConstantArb, ABArb, ABArb]
```

And you'd be testing for a much larger set of edge-cases.

Comment (Avaq) [reply test/index.js:331]: > `ConstantArb(Constant)(tie('AnyConstantArb'))`

I see the problem here.

Comment (Avaq) [reply test/index.js:331]: To fix that, I guess we'd have to define some kind of `TypeRepWithArbPairArb` (an arbitrary pair like `[String, jsc.string]`, for example) to generate valid ConstantArbs to be used in recursive definitions of AnyConstantArb. Maybe a little over-ambitious for now. 

I took the approach above from recent changes I made to the Fluture tests, and I figured they could be applied to other types as well. `Constant` is an odd exception because it's a specialized type by nature.

Comment (davidchambers) [reply test/index.js:331]: I love this idea, Aldwin. Although it's not easy to apply here, we should certainly adopt it for our other ADTs. If you're interested in working on this, I'd love to see a pull request for one of the other projects. :D

File: index.js (hunk 3 above)
Comment (Avaq) [reply index.js]: Has an issue been posted in sanctuary-type-classes to keep track of this?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+/*  |  |  | |  |  | |  |  | |  |  |   |  |   |  |  | |  |  |   |  |    */
+/*  |_____| |_____| |__|__| |_____|   |__|   |__|__| |__|__|   |__|    */
+
+//. <a href="https://github.com/fantasyland/fantasy-land"><img alt="Fantasy Land" src="https://raw.githubusercontent.com/fantasyland/fantasy-land/master/logo.png" width="75" height="75" align="left"></a>
+//.
+//. # sanctuary-constant
+//.
+//. A value of type `Constant a b` always contains exactly one value,
+//. of type `a`. Mapping over a `Constant a b` has no effect because
+//. the `b -> c` function is never applied.
```
Comment (Avaq) [inline index.js:14]: I don't really understand the Constant type, why it takes a type representative, and how that type representative relates to Constant's the type parameters. If you care to explain it to me, maybe that explanation can be added here to the introduction. :)

**Review by Avaq — APPROVED** [review]

File: index.js (hunk 3 above)
Comment (davidchambers) [reply index.js]: > Then if you have some function that can take a value and obtain [its] type rep, and __a function that can take a type rep and determine whether it implements a typeclass__, their composition is `test`.

fantasyland/fantasy-land#315 is the ultimate solution to this problem, I believe. Implementing that proposal is near the top of my list of Sanctuary-related tasks. :)

File: index.js (hunk 4 above)
Comment (davidchambers) [reply index.js:14]: In Haskell, we can do this:

```haskell
> import Data.Functor.Constant

> pure 42 :: Constant String Int
Constant ""
```

We provided `42 :: Int` in “`b`” position; `"" :: String` materialized in “`a`” position.

In JavaScript, keeping track of `a` and `b` in the type system is not an option so we must provide type information explicitly:

```javascript
> S.of (Constant (String)) (42)
Constant (String) ('')
```

File: index.js [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+  //. . 'Extend          ❌   ',
+  //. . 'Comonad         ❌   ',
+  //. . 'Contravariant   ❌   ' ]
+  //. ```
+
+  //# Constant :: TypeRep a -> a -> Constant a b
+  //.
+  //. Constant's sole data constructor. The [type representative][] makes
+  //. `Constant (M)` an [Applicative][]-compatible type representative if
+  //. `M` represents some monoidal type.
```
Comment (davidchambers) [inline index.js]: Is this helpful, @Avaq?

File: index.js (hunk 4 above)
Comment (Avaq) [reply index.js:14]: > `"" :: String` materialized

Woah. o.o .... Ahh, because String is a Monoid. So `Constant (Boolean)` will throw when instantiated.

Alright. That makes things clearer. I don't really see how this would be used in practice, and that's probably what most of my confusion stems from.

File: index.js (hunk 5 above)
Comment (Avaq) [reply index.js]: I read this a few times yesterday, but without being able to think up a practical use-case, I couldn't understand what it's for.

Comment (masaeedu) [reply index.js]: @Avaq Imagine you're trying to derive:

```
foldMap :: Monoid m -> (a -> m) -> f a -> m
```

from:

```
traverse :: Applicative g -> (a -> g b) -> f a -> g (f b)
```

So you start writing:

```
const foldMap = ({ empty, append }) => am => fa => {
  // ?
}
```

Now obviously the only tool you have available is `traverse`, from which you're trying to derive `foldMap`, so you write:

```
const foldMap = ({ empty, append }) => am => fa => traverse(/* ?1 */)(/* ?2 */)(/* ?3 */)
```

Now you're stuck with the task of producing an `Applicative g` to feed into `?1`, an `a -> g b` to feed into `?2` and a `f a` to feed into `?3`. It just so happens we already have an `fa :: f a` we received from the arguments, so that takes care of `?3` at least:

```
const foldMap = ({ empty, append }) => am => fa => traverse(/* ?1 */)(/* ?2 */)(fa)
```

Let's think about the types again:

```
traverse :: Applicative g -> (a -> g b) -> f a -> g (f b)
foldMap  :: Monoid m      -> (a -> m)   -> f a -> m
```

we can see a few things:

- it must be the case that `g (f b) = m`
- `g` (whatever it is) needs to have an instance `Applicative g`
- this derivation must work regardless of what traversable `f` is: i.e. `f b` could be any type whatsoever
- `m` again is beyond our control: so whatever `(a -> m)` the user provides to `foldMap`, it is precisely for that `m` that we must have `g (f b) = m`

So, what's an applicative functor `g` such that `g x = m` for an arbitrary type `x` and an arbitrary monoid `m`? :wink:

File: index.js [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      if (Z.Semigroup.test (value)) {
+        constant['fantasy-land/concat'] = Constant$prototype$concat;
+        constant['fantasy-land/ap'] = Constant$prototype$ap;
+      }
+      constant.value = value;
+      return constant;
+    }
+
+    //  XXX: sanctuary-show@3 only respects @@show on "objects". Fool it.
+    Constant$bound[Symbol.toStringTag] = 'Object';
```
Comment (davidchambers) [inline index.js]: @Avaq, here's a workaround for the issue we discussed not so long ago.

Comment (Avaq) [reply index.js]: Would this become a standard way of working? It seems a bit hacky.

Comment (davidchambers) [reply index.js]: sanctuary-js/sanctuary-show#28
